### PR TITLE
[Explorer]: add reference gas price calculation

### PR DESCRIPTION
**Pull request by plam-ml**
## Description 

- Add reference gas price for next epoch
- Update next epoch card
- https://mysten.atlassian.net/browse/APPS-808

Screenshot 2023-04-20 at 2 26 39 PM

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: apps/explorer/src/pages/epochs/EpochDetail.tsx [hunk 1]
```diff
@@ -36,6 +38,41 @@ function SuiStats({
...
 
+function getRefGasPrice(validators: SuiValidatorSummary[] | undefined) {
+    if (!validators?.length) {
+        return 0;
+    }
+
+    const REF_THRESHOLD = 66.67;
+
+    const sortedByGasPrice = [...validators].sort(
+        (a, b) => Number(a.gasPrice) - Number(b.gasPrice)
```
Comment (Jordan-Mysten) [inline apps/explorer/src/pages/epochs/EpochDetail.tsx:49]: Should this be bigint instead of number?

File: apps/explorer/src/pages/epochs/EpochDetail.tsx [hunk 2]
```diff
@@ -36,6 +38,41 @@ function SuiStats({
...
+    );
+    const totalStaked = validators.reduce(
+        (acc, cur) => acc + Number(cur.stakingPoolSuiBalance),
```
Comment (Jordan-Mysten) [inline apps/explorer/src/pages/epochs/EpochDetail.tsx:52]: Same here, bigint

File: apps/explorer/src/pages/epochs/EpochDetail.tsx [hunk 3]
```diff
@@ -36,6 +38,41 @@ function SuiStats({
     );
 }
 
+function getRefGasPrice(validators: SuiValidatorSummary[] | undefined) {
```
Comment (Jordan-Mysten) [inline apps/explorer/src/pages/epochs/EpochDetail.tsx:41]: Can just make the arg optional like `validators?: SuiValidatorSummary[]`

File: apps/explorer/src/pages/epochs/EpochDetail.tsx [hunk 4]
```diff
@@ -36,6 +38,41 @@ function SuiStats({
...
+    if (!validators?.length) {
+        return 0;
+    }
+
+    const REF_THRESHOLD = 66.67;
```
Comment (Jordan-Mysten) [inline apps/explorer/src/pages/epochs/EpochDetail.tsx:46]: Define this outside of this function

File: apps/explorer/src/pages/epochs/EpochDetail.tsx [hunk 5]
```diff
@@ -36,6 +38,41 @@ function SuiStats({
...
+
+    while (sumOfStakes <= REF_THRESHOLD) {
+        const validator = sortedByGasPrice.shift();
+        if (!validator) {
+            break;
+        }
+
+        const currentGasPrice = Number(validator.gasPrice);
+        const stake = Number(validator?.stakingPoolSuiBalance);
+        const stakeShare = (stake / totalStaked) * 100;
```
Comment (Jordan-Mysten) [inline apps/explorer/src/pages/epochs/EpochDetail.tsx:67]: All of these should be bigints

**Review by Jordan-Mysten — CHANGES_REQUESTED** [review]

File: apps/explorer/src/ui/RingChart.tsx [hunk 6]
```diff
@@ -17,33 +18,31 @@ export interface RingChartProps {
...
                 {title}
             </Heading>
             <div className="flex flex-col items-start justify-center gap-2">
-                {data.map(({ color, label }) => (
+                {data.map(({ color, label, value }) => (
                     <div className="flex items-center gap-1.5" key={label}>
                         <div
                             style={{ backgroundColor: color }}
-                            className="h-3 w-3 rounded-sm"
+                            className={clsx('h-3 w-3 rounded-sm')}
```
Comment (Jordan-Mysten) [inline apps/explorer/src/ui/RingChart.tsx:30]: Why this change?

File: apps/explorer/src/pages/epochs/EpochDetail.tsx (hunk 5 above)
Comment (plam-ml) [reply apps/explorer/src/pages/epochs/EpochDetail.tsx]: Moved this code out into a util in `core` and updated to use `BigInt` conversions.

**Review by mystie711 — COMMENTED** [review]
Please update the rounded corner values in the card. 
Other than that looks great!

**Comment by mystie711** [comment]
I also noticed that the blue circle of active validators is not using a gradient fill. could we please?

**Comment by mystie711** [comment]
Also, one last point...if there are no new validators this round, let's hide that item instead of showing "0 New"

**Review by mystie711 — APPROVED** [review]

File: apps/core/src/utils/getRefGasPrice.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Util to get the Reference Gas Price from a list of validators
+ * 1. Sort validators by gas price
+ * 2. Calculate the stake threshold at 66.67% of total stake amount
+ * 3. Add up the stake amount of each validator until the threshold is reached
+ * 4. Return the gas price of the last validator that was added to the sum
+ */
+export function getRefGasPrice(validators?: SuiValidatorSummary[]) {
+    if (!validators?.length) {
+        return '0';
```
Comment (Jordan-Mysten) [inline apps/core/src/utils/getRefGasPrice.ts:17]: Why a string?

File: apps/core/src/utils/getRefGasPrice.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+const REF_THRESHOLD = 66.67;
+
+/**
+ * Util to get the Reference Gas Price from a list of validators
+ * 1. Sort validators by gas price
+ * 2. Calculate the stake threshold at 66.67% of total stake amount
+ * 3. Add up the stake amount of each validator until the threshold is reached
+ * 4. Return the gas price of the last validator that was added to the sum
+ */
+export function getRefGasPrice(validators?: SuiValidatorSummary[]) {
```
Comment (Jordan-Mysten) [inline apps/core/src/utils/getRefGasPrice.ts:15]: seems weird to me that this returns a string, I feel like a bigint is a better return value

File: apps/explorer/src/ui/RingChart.tsx [hunk 9]
```diff
@@ -17,81 +19,97 @@ export interface RingChartProps {
...
 
     return (
-        <div className="grid grid-flow-row items-center justify-start xl:grid-cols-2 xl:justify-center">
-            <div className="self-start">
-                <Legend data={data} title={title} />
-            </div>
-
-            <div className="relative">
+        <div className="flex items-center gap-5">
+            <div className="relative h-24 min-h-[96px] w-24 min-w-[96px]">
```
Comment (Jordan-Mysten) [inline apps/explorer/src/ui/RingChart.tsx:97]: Why an explicit height/width and also a min one?

Comment (Jordan-Mysten) [reply apps/explorer/src/ui/RingChart.tsx:97]: Also I never love explicit heights and widths, especially in a base component like this. We either should make this more generic or we should hoist the entire component out of here and into the relevant UI surface

File: apps/explorer/src/pages/epochs/stats/ValidatorStatus.tsx [hunk 10]
```diff
@@ -7,20 +7,32 @@ import { RingChart } from '~/ui/RingChart';
...
         <RingChart
-            title="Validators in Next Epoch"
+            title={`Validators in Epoch ${nextEpoch}`}
             suffix="validators"
             data={[
                 {
                     value: data.activeValidators.length,
                     label: 'Active',
-                    color: '#589AEA',
+                    color: `linear-gradient(315deg, ${activeGradient[0].color} ${activeGradient[0].percent}, ${activeGradient[1].color} ${activeGradient[1].percent})`,
```
Comment (Jordan-Mysten) [inline apps/explorer/src/pages/epochs/stats/ValidatorStatus.tsx:26]: both color and gradient are needed?

File: apps/explorer/src/ui/RingChart/RingChart.tsx [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    label: string;
+    color?: string;
+    gradient?: Gradient;
+}[];
+
+export interface RingChartProps {
+    data: RingChartData;
+}
+
+export function RingChart({ data }: RingChartProps) {
```
Comment (plam-ml) [inline apps/explorer/src/ui/RingChart/RingChart.tsx:22]: Mostly copy-pasted from existing code

File: apps/explorer/src/ui/RingChart/RingChartLegend.tsx [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    const valuesMap = values.map(
+        ({ percent, color }) => `${color} ${percent}%`
+    );
+
+    gradientResult.push(...valuesMap);
+
+    return `linear-gradient(${gradientResult.join(',')})`;
+}
+
+export function RingChartLegend({ data, title }: RingChartLegendProps) {
```
Comment (plam-ml) [inline apps/explorer/src/ui/RingChart/RingChartLegend.tsx:31]: Mostly copy-pasted from existing code

File: apps/explorer/src/ui/RingChart/index.ts [hunk 13]
```diff
@@ -0,0 +1,5 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+export { RingChart as default, type RingChartProps } from './RingChart';
```
Comment (Jordan-Mysten) [inline apps/explorer/src/ui/RingChart/index.ts:4]: Can you just move these to index?

Comment (Jordan-Mysten) [reply apps/explorer/src/ui/RingChart/index.ts:4]: Also prefer named export.

File: apps/explorer/src/pages/epochs/stats/ValidatorStatus.tsx [hunk 14]
```diff
@@ -1,33 +1,82 @@
...
+
+                <div className="self-start">
+                    <RingChartLegend
+                        data={chartData}
+                        title={`Validators in Epoch ${nextEpoch}`}
+                    />
+                </div>
+            </div>
+
+            <div className="mt-8 flex items-center justify-between rounded-lg border border-solid border-steel px-3 py-2">
```
Comment (Jordan-Mysten) [inline apps/explorer/src/pages/epochs/stats/ValidatorStatus.tsx:62]: mt + py can be a little tricky, worth double checking the numbers

File: apps/explorer/src/ui/RingChart/index.ts (hunk 13 above)
Comment (plam-ml) [reply apps/explorer/src/ui/RingChart/index.ts]: Updated. If we move this into `index`, then we will get 2 different lines for importing `RingChart` and `RingChartLegend` which are from the same folder. Though we could move them into 1 big file but that would be a 200 lines file, unless you think it is better to have everything in 1 file instead of index exporting.

File: apps/explorer/src/pages/epochs/stats/ValidatorStatus.tsx (hunk 14 above)
Comment (plam-ml) [reply apps/explorer/src/pages/epochs/stats/ValidatorStatus.tsx:62]: The padding is for the border, margin top is for the upper content.

**Review by Nikhil-Mysten — CHANGES_REQUESTED** [review]
not really seeing anything that sticks out - does this need to be behind a feature flag for main net launch?

**Review by Nikhil-Mysten — APPROVED** [review]

**Comment by plam-ml** [comment]
@Nikhil-Mysten nope, we can release this. This has been an ask for awhile

File: apps/explorer/src/ui/RingChart/index.ts (hunk 13 above)
Comment (Jordan-Mysten) [reply apps/explorer/src/ui/RingChart/index.ts]: 200 lines doesn't scare me

File: apps/explorer/src/ui/Card.tsx [hunk 15]
```diff
@@ -10,10 +10,12 @@ const cardStyles = cva(null, {
             default: 'bg-gray-40',
             highlight: 'bg-success-light',
             lightBlue: 'bg-sui/10',
+            white: 'bg-white border border-steel',
```
Comment (Jordan-Mysten) [inline apps/explorer/src/ui/Card.tsx:13]: I feel like I've seen other in-flight PRs add this but with different styles.

**Review by Jordan-Mysten — APPROVED** [review]

File: apps/explorer/src/ui/Card.tsx (hunk 15 above)
Comment (plam-ml) [reply apps/explorer/src/ui/Card.tsx:13]: I just took around to other PRs currently, saw @williamrobertson13 added some styling on `Card` but not to the bg color. I will merge this so others can rebase if there are pending changes related.
